### PR TITLE
Collection Wait Handler Implementation [14488]

### PR DESCRIPTION
**Pull request by jparisu**
After #203

### CODE REVIEW

**Comment by jparisu** [comment]
This implementation requires to have a more specific classes for:
- Collections with internal thread safe methods (as DBQueue)
- Collections without internal thread safe methods (as std::queue)

- [ ] TODO

File: ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+/**
+ * Check that pushing and popping values works as expected from the same thread
+ * Using std::string as object to use inside collection.
+ *
+ * CASES:
+ * - Push and pop one value by moving it
+ *
+ * TODO:
+ * This test will not work as DBQueue do not allow moving values, it only copy them
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp:80]:
```suggestion
 * This test will not work as DBQueue do not allow moving values, it only copies them
```

File: ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+    std::string lvalue("test_data");
+
+    handler.add_value(lvalue);
+
+    // Getting first value
+    std::string pop_value = handler.wait_next_value();
+    EXPECT_EQ(lvalue, pop_value);
+
+    // They should be same object, check that modifying one modifies the other
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp:120]:
```suggestion
    // They should be different objects, check that modifying one does not modify the other
```

File: ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        });
+
+        thread_A.join();
+        thread_B.join();
+        thread_C.join();
+
+        ASSERT_TRUE(popped_1 && popped_2);
+    }
+
+    {
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp:185]:
```suggestion
    // 2
    {
```

File: ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        std::thread thread_C([&]() {
+            handler.add_value(1);
+            handler.add_value(2);
+            handler.add_value(3);
+        });
+
+        thread_C.join();
+
+        // Wait for notification to arrive to waiting threads
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp:248]: Use time constant defined above?

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: ddsrouter_event/include/ddsrouter_event/wait/CollectionWaitHandler.hpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    using CounterWaitHandler::enabled;
+    using CounterWaitHandler::stop_and_continue;
+
+    /////
+    // Add values methods
+
+    /**
+     * @brief Add a new value to the collection. Use move constructor.
+     *
+     * This method will awake ONE thread waiting for data to be available if there is any waiting
```
[the remark this answers is not in the document]
Comment (jparisu) [reply ddsrouter_event/include/ddsrouter_event/wait/CollectionWaitHandler.hpp:60]: We already studied this and in the future we will reimplement this class without CoutnerWaitHandler.
However, this comment references the behaviour of the method, not its implementation. Thus, it is true that will only awake a thread, while others will still be waiting in wait (after awaken and check the predicate is false).

File: ddsrouter_event/include/ddsrouter_event/wait/impl/ConsumerWaitHandler.ipp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+namespace eprosima {
+namespace ddsrouter {
+namespace event {
+
+template <typename T>
+ConsumerWaitHandler<T>::ConsumerWaitHandler(
+        bool enabled /* = true */)
+    : WaitHandler(0, enabled)
+{
+    logDebug(DDSROUTER_WAIT_COLLECTION, "Created Consumer Wait Handler with type " << TYPE_NAME(T) << ".");
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/include/ddsrouter_event/wait/impl/ConsumerWaitHandler.ipp:34]: Update log catergory names.

File: ddsrouter_event/include/ddsrouter_event/wait/impl/DBQueueWaitHandler.ipp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+#define _DDSROUTEREVENT_WAIT_IMPL_DBQUEUEWAITHANDLER_IPP_
+
+namespace eprosima {
+namespace ddsrouter {
+namespace event {
+
+template <typename T>
+void DBQueueWaitHandler<T>::add_value_(T&& value)
+{
+    logDebug(DDSROUTER_WAIT_COLLECTION, "Moving element to DBQueue.");
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/include/ddsrouter_event/wait/impl/DBQueueWaitHandler.ipp:30]: Same here.

File: ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+ * Check that pushing and popping values works as expected from the same thread
+ * Using std::string as object to use inside collection.
+ *
+ * CASES:
+ * - Push and pop one value by moving it
+ *
+ * TODO:
+ * This test will not work as DBQueue do not allow moving values, it only copies them
+ */
+TEST(DBQueueWaitHandlerTest, push_pop_one_thread_string_move)
```
Comment (juanlofer-eprosima) [inline ddsrouter_event/test/unittest/wait/DBQueueWaitHandlerTest.cpp:82]: This test might not do what is intended.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**